File: library/src/main/java/io/github/anderscheow/validator/rules/common/FutureRule.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;

import java.text.DateFormat;
import java.text.ParseException;
import java.util.Date;

import io.github.anderscheow.validator.rules.BaseRule;

public class FutureRule extends BaseRule {

    private DateFormat dateFormat;

    public FutureRule(@Nullable DateFormat dateFormat) {
        super("Does not match future rule");
        this.dateFormat = dateFormat;
    }

    public FutureRule(@Nullable DateFormat dateFormat, @StringRes int errorRes) {
        super(errorRes);
        this.dateFormat = dateFormat;
    }

    public FutureRule(@Nullable DateFormat dateFormat, @NonNull String errorMessage) {
        super(errorMessage);
        this.dateFormat = dateFormat;
    }

    @Override
    public boolean validate(Object value) {
        if (value == null) {
            throw new NullPointerException();
        }

        Date parsedDate = null;

        if (value instanceof String) {
            if (dateFormat == null) {
                throw new NullPointerException();
            }

            try {
                parsedDate = dateFormat.parse((String) value);
            } catch (ParseException ignored) {
            }
        } else if (value instanceof Date) {
            parsedDate = (Date) value;
        } else {
            throw new ClassCastException("Required String or Date value");
        }

        return parsedDate != null && parsedDate.after(new Date());
    }
}

File: library/src/main/java/io/github/anderscheow/validator/rules/common/FutureRule.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import io.github.anderscheow.validator.rules.Rule
import java.text.DateFormat
import java.text.ParseException
import java.util.*

class FutureRule : Rule {

    private var dateFormat: DateFormat

    constructor(dateFormat: DateFormat, @StringRes errorRes: Int) :
            super(errorRes) {
        this.dateFormat = dateFormat
    }

    constructor(dateFormat: DateFormat, errorMessage: String) :
            super(errorMessage) {
        this.dateFormat = dateFormat
    }

    override fun validate(value: String?): Boolean {
        if (value == null) {
            throw NullPointerException()
        } else {
            var parsedDate: Date? = null

            try {
                parsedDate = dateFormat.parse(value)
            } catch (ignored: ParseException) {
            }

            return parsedDate != null && parsedDate.after(Date())
        }
    }
}

fun future(dateFormat: DateFormat, @StringRes errorRes: Int): FutureRule =
    FutureRule(dateFormat, errorRes)

fun future(dateFormat: DateFormat, errorMessage: String): FutureRule =
    FutureRule(dateFormat, errorMessage)